### DISCUSSION TITLE
Making Library more Testing and TDD friendly

### DISCUSSION THREAD

**Discussion by sspates** [nats-io/nats.net#311] [Ideas]
This discussion is to bring together ideas around making the v2 library more Testing and TDD friendly as the interfaces as they currently exist are unusable because other elements of the system (e.g. JetstreamContext) can't use the interfaces. 

Some PRs/Issues related:
https://github.com/nats-io/nats.net.v2/pull/297
https://github.com/nats-io/nats.net.v2/issues/301
https://github.com/nats-io/nats.net.v2/issues/300
https://github.com/nats-io/nats.net.v2/issues/299

**Comment by sspates** [comment]
@mtmk @caleblloyd @jasper-d @to11mtm

Starting discussion around the "interfaces" for the library as there seems to be a lot of discussion that really is design of the library and not just specific to one PR or another.   

Some of the details that probably need to be worked out from the Synadia side before PRs can be adjusted or put place could be:
- Interfaces or abstracts
- Style of Eventing (event, delegates, func/action)
- virtual, internal or protected methods
- obsoletion or contract breaks to enable the public "interfaces" and disable non-public elements
- GetStats() vs .Net Metrics

**Reply by mtmk** [reply]
I would like us to come to a decision as a community especially around specific language related issues.

> Interfaces or abstracts

I'd say interfaces. @to11mtm made an interesting suggestion but we can work on that separately I think.

> Style of Eventing (event, delegates, func/action)

Sounds like we'd be going with async delegates. if @mnmr doesn't have spare time, I can jump in there.

> virtual, internal or protected methods
> obsoletion or contract breaks to enable the public "interfaces" and disable non-public elements

We can be somewhat flexible here as it's still early days. I'd say if we can provide a easy fix at the cost of recompiling we should be able to release a few more breaking realises. We can take it case by case.

> GetStats() vs .Net Metrics

in the long run I'd like to see Metrics implemented.

**Comment by mtmk** [comment]
I feel we're reaching a consensus gradually. We should be able to make some progress with #297 and merge it for example. I'll add a few comments to clarify some points as I think how we as community want to move forward.

**Comment by sspates** [comment]
@mtmk Just added the updates you requested to my PR.  

If the idea is to use interfaces everywhere, then how are they going to be passed into the other elements like JetstreamContext and still provide working code. Is the plan to create factories that return interfaces instead of the existing public constructors or something more like the v1 client CreateJetStreamContext methods on the IConnection.

**Reply by mtmk** [reply]
What would you prefer? If JetstreamContext can take the interface that would be the simplest I guess. I think we were using a few internal methods within JetStream need to check them first.

**Reply by caleblloyd** [reply]
`NATS.Client.Core` does not depend on `NATS.Client.JetStream` so if we did `INatsConnection.CreateJetStreamContext` it would need to be via an extension method in `NATS.Client.JetStream`

**Reply by mtmk** [reply]
We can provide factories in `NATS.Client.Hosting` (would need to add dependencies) and/or `NATS.Net`.

**Reply by sspates** [reply]
You could add factories in the NATS.Client.Jetstream to construct contexts like this:
`public interface INatsJSContextFactory {
    INatsJSContext CreateContext(INatsConnection connection, NatsJSOpts opts);
}
`

Then there would be an explicit implementation:
`public interface NatsJSContextFactory {
    INatsJSContext CreateContext(INatsConnection connection, NatsJSOpts opts)
    {
        return new NatsJSContext((NatsConnection)connection, opts);
    }
}
` 

The same strategy could also be used in NATS.Client.KeyValueStore and NATS.Client.ObjectStore.

**Reply by mtmk** [reply]
That sounds reasonable to me.

**Reply by sspates** [reply]
@mtmk I will put in a PR with that adjustment in place.

**Reply by sspates** [reply]
@mtmk @caleblloyd : https://github.com/nats-io/nats.net.v2/pull/330

**Comment by mtmk** [comment]
I think we have addressed most of the issues with [release 2.1.0-preview.4](https://github.com/nats-io/nats.net.v2/releases/tag/v2.1.0-preview.4) except the following to my knowledge:
* #301 which has an open question at the moment - feel free to chip in
* #348 Question about event dispatch and whether we should still support sync events as well as async
* ~~We need to add the new async events to the interface (not tracked anywhere)~~ #353

**Reply by sspates** [reply]
@mtmk @caleblloyd One last PR to add the Async events to the interface: https://github.com/nats-io/nats.net.v2/pull/353

**Reply by mtmk** [reply]
thank you @sspates it's released with 2.1.0-preview.5.

**Comment by mtmk** [comment]
Thanks everyone for your suggestions and contributions especially @sspates 💯 closing this as I think we've implemented all the suggestions here. feel free to reopen if you think there is more to discuss or if I left anything out.
